### PR TITLE
Make text bubble width different from cell width

### LINKED ISSUE #1112: Bottom bubble shifts on rotation, scroll position not maintained. Make bubbles fixed-width?
@jessesquires 

In the demo app, when you rotate the view the content offset does not land very close to the bubbles that  were visible before rotation.
screen shot 2015-07-23 at 3 11 05 pm
screen shot 2015-07-23 at 3 10 54 pm

Part of the problem is that cells have a different size in landscape vs portrait.
Chat bubbles widths are controlled by JSQMessagesCollectionViewFlowLayout.itemWidth.
I've tried subclassing JSQMessagesCollectionViewFlowLayout and returning a constant for itemWidth (regardless of landscape or portrait) but in landscape, all of my bubbles appear in the center of the view.

It turns out that itemWidth is used to control the overall width of the cell, as well as the maximum width of the bubble portion of the cell.

Would you be opposed to a PR which splits the ways in which itemWidth is used?
I'd like to introduce a new property: textBubbleWidth
For the base class (JSQMessagesCollectionViewFlowLayout), textBubbleWidth will just call itemWidth
````
- (CGFloat)textBubbleWidth {
	return [self itemWidth];
}
````
But in subclasses, we can return the narrower of the width or the height, and in this way, we can make the rows the same size/height regardless of landscape or portrait. This will go a long way toward getting the computation of contentOffset correct.

Ray

:muscle: :sunglasses: :facepunch:

### PR DESCRIPTION
**Pull request by walsh2000**
create an override point for subclasses
    
break apart cell-width from text bubble width, allowing subclasses to have different values for each

This is necessary for consumers to deal with #1112, though it does not resolve #1112.

### CODE REVIEW

**Comment by walsh2000** [comment]
@jessesquires 

I don't know how my PR might have caused this build failure. It is a brain-dead safe commit.
It looks like a few other PRs are failing too. Is there something wrong with the project or perhaps is there a better branch to submit against?

**Comment by jessesquires** [comment]
@walsh2000 - failure was unrelated to your changes. some other things are borked at the moment.

**Comment by jessesquires** [comment]
Just pushed fixes to `develop`, try merging from upstream

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m [hunk 1]
```diff
@@ -48,6 +48,7 @@ @interface JSQMessagesCollectionViewFlowLayout ()
 @property (assign, nonatomic) CGFloat latestDelta;
 
 @property (assign, nonatomic, readonly) NSUInteger bubbleImageAssetWidth;
+@property (assign, nonatomic) NSUInteger rotationIndependentLayoutWidth;
```
Comment (jessesquires) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: Y U NO `BOOL` ?

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h [hunk 2]
```diff
@@ -75,6 +75,26 @@ FOUNDATION_EXPORT const CGFloat kJSQMessagesCollectionViewAvatarSizeDefault;
...
+ *  that when you rotate from landscape <-> portrait, the same area of the conversation is on screen.
+ *  This is the behavior you can see in the iOS Messages app.
+ *
+ *  When the value it set to `NO` text bubbles change width/height significantly, depending upon bubble-text
+ *  length, resulting in a very different total scroll height, which means that when you rotate
+ *  from landscape <-> portrait, a different portion of the conversation is visible.
+ *
+ *  @warning Though this feature is mostly stable, it is still considered an experimental feature.
+ */
+@property (assign, nonatomic) BOOL rotationIndependentLayout;
```
Comment (jessesquires) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h]: naming: how about `usesFixedWidthMessageBubbles`?

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h [hunk 3]
```diff
@@ -75,6 +75,26 @@ FOUNDATION_EXPORT const CGFloat kJSQMessagesCollectionViewAvatarSizeDefault;
...
+ *  When the set to `YES`, there will be a large gap between bubbles down the center.
+ *  However, this has the benefit of making the total scroll height of the layout the same, which means
+ *  that when you rotate from landscape <-> portrait, the same area of the conversation is on screen.
+ *  This is the behavior you can see in the iOS Messages app.
+ *
+ *  When the value it set to `NO` text bubbles change width/height significantly, depending upon bubble-text
+ *  length, resulting in a very different total scroll height, which means that when you rotate
+ *  from landscape <-> portrait, a different portion of the conversation is visible.
+ *
+ *  @warning Though this feature is mostly stable, it is still considered an experimental feature.
```
Comment (jessesquires) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h]: :scream: :fearful: experimental? 

what issues are there?

**Comment by jessesquires** [comment]
Hey @walsh2000 -- So, I actually have some great changes planned that will add a hook in the `FlowLayout` object to allow clients to calculate their own bubble sizes. 

My intent with this is to finally "fix" #347 — if using a custom font, you can use this hook to resize your bubbles. (And add your own magic numbers :smirk: ) This is now very relevant to this PR.

I'll push those changes asap. You could then hook into this mechanism.

(Thinking now if I should merge first, or if you should. Either way there will be conflicts to resolve.)

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m (hunk 1 above)
Comment (walsh2000) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.m]: Hah! I didn't notice that. I must have templated off of the line above.

File: JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h (hunk 3 above)
Comment (walsh2000) [inline JSQMessagesViewController/Layout/JSQMessagesCollectionViewFlowLayout.h]: copy/paste typo.

**Comment by walsh2000** [comment]
@jessesquires 
I'll handle the marge conflicts, or we could scrap this whole PR and I'll come at it with whatever the new way requires. This wasn't a challenging change once the problem was understood.

**Comment by walsh2000** [comment]
Bah.. i just noticed travis failing. I don't think these changes caused this exception. It looks like simulator compilation issues.

**Comment by jessesquires** [comment]
Hey @walsh2000 - I finally pushed my changes. See 983fb58 and https://github.com/jessesquires/JSQMessagesViewController/issues/347#issuecomment-130533179 for details.

**Comment by walsh2000** [comment]
@jessesquires 

I messed up a little with that merge, but eventually got it right. Let me know if you'd like me to start over.
I didn't change much though... the default bubble size calculator still requests the text size from the layout.
I could have written a separate bubble size calculator, and I will if you don't like this direction.

**Comment by jessesquires** [comment]
@walsh2000 --- Hm... yeah, looks like something went wrong with that merge. Maybe close this, and open a different PR?

>I could have written a separate bubble size calculator, and I will if you don't like this direction.

So I'm thinking that *everything* can happen inside of `JSQMessagesBubbleSizeCalculator`. We can move the `usesFixedWidthMessageBubbles` property there (as read-only) and add it to the initializer. From there, tweak the sizing logic based on the value of `usesFixedWidthMessageBubbles`. How's that sound?

**Comment by jessesquires** [comment]
If you need more information from the layout inside the BubbleSizeCalculator, we can pass more parameters to the delegate method.

**Comment by walsh2000** [comment]
@jessesquires 
Yea, that sounds good. I'll work that up and submit a new PR

**Comment by jessesquires** [comment]
Thanks @walsh2000 ! :smile:

**Comment by walsh2000** [comment]
I submitted #1164 to replace this PR

**Comment by jessesquires** [comment]
:sparkles: :bow: :sparkles: such excellent book-keeping! :+1:

**Pull request closed without merging**